runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/EngineConfiguration.java

```diff
@@ -385,17 +385,32 @@ private static int defaultWorkersCapacity(
 
         final int slotCapacity = ENGINE_BUFFER_SLOT_CAPACITY.get(config);
         final double percentMemory = ENGINE_MEMORY_PERCENTAGE.get(config);
+        final int totalEventsBufferCapacity = ENGINE_EVENTS_BUFFER_CAPACITY.get(config) * numberOfCores;
 
-        long maxAllowedForBuffers = (long) (percentMemory * totalMemorySize);
+        long maxAllowedForMemory = (long) (percentMemory * totalMemorySize) - totalEventsBufferCapacity;
+
+        Path directory = Paths.get(ENGINE_DIRECTORY.get(config));
+        long usableDiskSpace = directory.toFile().getUsableSpace();
+
+        long maxAllowedForBuffers = Math.min(maxAllowedForMemory, usableDiskSpace);
 
         // Streams + Pool
         long bufferCapacity = slotCapacity + slotCapacity;
-        long eventsBufferCapacity = ENGINE_EVENTS_BUFFER_CAPACITY.get(config);
-        long budgetBufferCapacity = BudgetsLayout.SIZEOF_BUDGET_ENTRY * 512L;
-        long totalBufferCapacity = numberOfCores * (bufferCapacity + budgetBufferCapacity + eventsBufferCapacity);
-        int newWorkersCapacity = (int) (maxAllowedForBuffers / totalBufferCapacity);
+        long budgetBufferCapacity = BudgetsLayout.SIZEOF_BUDGET_ENTRY;
+        long totalBufferCapacity = numberOfCores * (bufferCapacity + budgetBufferCapacity);
```

Comment:
I realized that although we default `workers` to `available processors`, the calculation should actually be based on `workers`, not `numberOfCores`.

So, we actually only need `osBean` for `getTotalMemorySize()`.

For workers, we would do this instead.
```
int workers = ENGINE_WORKERS.getAsInt(config);
```

Also, for clarity suggest we do the following:
```
int workerBufferCapacity = slotCapacity + slotCapacity + budgetBufferCapacity;
int totalBufferCapacity = workerBufferCapacity * workers;
```